static/js/compose_ui.js

```diff
@@ -372,23 +405,199 @@ export function format_text(textarea, type) {
+        range.start >= 1 &&
+        text.length - range.end >= 3 &&
+        text.slice(range.start - 1, range.start) === "[" &&
+        text.slice(range.end, range.end + 2) === "](" &&
+        text.includes(")") &&
+        text.indexOf(")") >= range.end + 2;
+
+    const is_selection_link = () =>
+        range.length >= 4 &&
+        text[range.start] === "[" &&
+        text[range.end - 1] === ")" &&
+        text.includes("](") &&
+        text.indexOf("](") > range.start &&
+        text.indexOf("](") < range.end - 2;
+
+    if (is_selection_url()) {
+        const beginning = text.lastIndexOf("[", range.start);
+        const url = selected_text === "url" ? "" : " " + selected_text;
+        text =
+            text.slice(0, beginning) +
+            text.slice(beginning + 1, text.indexOf("]", beginning)) +
+            url +
+            text.slice(range.end + 1, text.length);
+
+        set(field, text);
+        field.setSelectionRange(beginning, beginning);
+        return;
+    } else if (is_selection_description_of_link()) {
+        let url = text.slice(range.end + 2, text.indexOf(")", range.end));
+        url = url === "url" ? "" : " " + url;
+        text =
+            text.slice(0, range.start - 1) +
+            text.slice(range.start, range.end) +
+            url +
+            text.slice(text.indexOf(")", range.end) + 1, text.length);
+        set(field, text);
+        field.setSelectionRange(range.start - 1, range.start - 1);
+        return;
+    } else if (is_selection_link()) {
+        const description = selected_text.split("](")[0].replace("[", "");
+        let url = selected_text.split("](")[1].replace(")", "");
+        url = url === "url" ? "" : " " + url;
+        text = text.slice(0, range.start) + description + url + text.slice(range.end, text.length);
+        set(field, text);
+        field.setSelectionRange(range.start - 1, range.start - 1);
+        return;
+    }
+
+    // Otherwise, we don't have code syntax, so we add it.
+    wrapSelection(field, link_syntax_start, link_syntax_end);
+
+    // Change selected text to `url` part of the syntax.
+    // If <text> marks the selected region, we're mapping:
+    // <text> => [text](<url>).
+    const new_start = range.end + "[](".length;
+    const new_end = new_start + "url".length;
+    field.setSelectionRange(new_start, new_end);
+}
+
+function add_spoiler_formatting(field, range, text, selected_text) {
+    let spoiler_syntax_start = "```spoiler \n";
+    let spoiler_syntax_end = "\n```";
+
+    const is_curser_at_header_position = () =>
+        range.length === 0 && text.slice(range.start - 11, range.start) === "```spoiler ";
+
+    const is_content_selected = () =>
+        text.slice(range.end, range.end + spoiler_syntax_end.length) === "\n```" &&
+        text.slice(range.start - 1, range.start) === "\n" &&
+        text.includes("```spoiler ") &&
+        text.lastIndexOf("```spoiler ") < range.start - 11;
```

Comment:
It would be very helpful for any complex blocks like this to have brief comments with with examples indicating the state that one is trying to capture.

Also, is it OK that some of these subtractions end up negative?

Reply:
It's fine that these subtractions can be negative, that just indicates that this check is not true